rfd/0144-client-tools-updates.md

```diff
@@ -0,0 +1,295 @@
+state: draft
+---
+
+# RFD 0144 - Client Tools Updates
+
+## Required Approvers
+
+* Engineering: @sclevine && @bernardjkim && @r0mant
+* Product: @klizhentas || @xinding33
+* Security: @reedloden
+
+## What/Why
+
+This RFD describes how client tools like `tsh` and `tctl` can be kept up to
+date, either using automatic updates or self-managed updates.
+
+Keeping client tools updated helps with security (fixes for known security
+vulnerabilities are pushed to endpoints), bugs (fixes for resolved issues are
+pushed to endpoints), and compatibility (users no longer have to learn and
+understand [Teleport component
+compatibility](https://goteleport.com/docs/upgrading/overview/#component-compatibility)
+rules).
+
+## Details
+
+### Summary
+
+Client tools like `tsh` and `tctl` will automatically download and install the
+required version for the Teleport cluster.
+
+Enrollment in automatic updates for client tools will be controlled at the
+cluster level. By default all clusters will be opted into automatic updates for
+client tools. Cluster administrators using MDM software like JamF will be able
+opt-out manually manage updates.
+
+Inspiration drawn from https://go.dev/doc/toolchain.
+
+### Implementation
+
+#### Client tools
+
+##### Automatic updates
+
+When `tsh login` is executed, client tools will check `/v1/webapi/ping` to
+determine if automatic updates are enabled. If the cluster's required version
+differs from the current binary, client tools will download and re-execute
+using the version required by the cluster.
+
+The original client tools binaries won't be overwritten. Instead, additional
+binaries will be downloaded and stored in `~/.tsh/bin` with `0555` permissions.
+A file named `version.yaml` in the same directory will maintain the mapping
+between proxies and tools versions, to support auto-purge and free up disk
+space from unused binaries.
+
+To enable concurrent operation of client tools, a locking mechanisms utilizing
+[syscall.Flock](https://pkg.go.dev/syscall#Flock) (for Linux and macOS) and
+[LockFileEx](https://learn.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-lockfileex)
+(for Windows) will be used.
+
+A side effect of automatic updates will be users will be able to connect to
+different Teleport clusters (running different versions of Teleport, including
+pre-releases) without juggling multiple versions of client tools.
+
+```
+$ tree ~/.tsh
+~/.tsh
+├── bin
+│  ├── 15.0.0
+│  │  ├── tctl
+│  │  └── tsh
+│  ├── 15.1.1
+│  │  ├── tctl
+│  │  └── tsh
+│  └── version.yaml
+├── current-profile
+├── keys
+│  └── proxy.example.com
+│     ├── cas
+│     │  └── example.com.pem
+│     ├── certs.pem
+│     ├── foo
+│     ├── foo-ssh
+│     │  └── example.com-cert.pub
+│     ├── foo-x509.pem
+│     └── foo.pub
+├── known_hosts
+└── proxy.example.com.yaml
+```
+
+Users can cancel client tools updates using `Ctrl-C`. This may be needed if the
```

Comment:
Users might expect `ctrl-c` to terminate `tsh` and return to a prompt, since most human-interfacing CLIs behave that way. We don't need to use a signal for this, since it's not interrupting execution -- could just do `(hit 'q' to skip)`

Reply:
How strong do you feel about this? `Ctrl-C` does mean interrupt so it feels reasonable to use. Will be easy to detect as well.

Reply:
When I've implemented non-terminating behaviors for `Ctrl-C`, I've noticed that users often hit it more than once, which can have unexpected results. It also requires de-registering and re-registering existing hooks. Reading `q` on stdin seems more straightforward to me.

That said, I have no preference at all -- `Ctrl-C` is definitely clear enough.

Reply:
Let's keep it at `Ctrl-C` for now, but I'll keep this in mind during implementation and share a few builds with you guys to see how it works out in real life.